src/Storages/RabbitMQ/StorageRabbitMQ.cpp

```diff
@@ -507,52 +511,15 @@ void StorageRabbitMQ::bindQueue(size_t queue_id, AMQP::TcpChannel & rabbit_chann
     /// AMQP::autodelete setting is not allowed, because in case of server restart there will be no consumers
     /// and deleting queues should not take place.
     rabbit_channel.declareQueue(queue_name, AMQP::durable, queue_settings).onSuccess(success_callback).onError(error_callback);
-    event_handler->startBlockingLoop();
-}
-
-
-bool StorageRabbitMQ::restoreConnection(bool reconnecting)
-{
-    size_t cnt_retries = 0;
-
-    if (reconnecting)
-    {
-        connection->close(); /// Connection might be unusable, but not closed
-
-        /* Connection is not closed immediately (firstly, all pending operations are completed, and then
-         * an AMQP closing-handshake is  performed). But cannot open a new connection until previous one is properly closed
-         */
-        while (!connection->closed() && cnt_retries++ != RETRIES_MAX)
-            event_handler->iterateLoop();
-
-        /// This will force immediate closure if not yet closed
-        if (!connection->closed())
-            connection->close(true);
-
-        LOG_TRACE(log, "Trying to restore connection to " + address);
-    }
-
-    auto amqp_address = connection_string.empty() ? AMQP::Address(parsed_address.first, parsed_address.second,
-                                                    AMQP::Login(login_password.first, login_password.second), vhost, secure)
-                                                  : AMQP::Address(connection_string);
-    connection = std::make_unique<AMQP::TcpConnection>(event_handler.get(), amqp_address);
-
-    cnt_retries = 0;
-    while (!connection->ready() && !stream_cancelled && cnt_retries++ != RETRIES_MAX)
-    {
-        event_handler->iterateLoop();
-        std::this_thread::sleep_for(std::chrono::milliseconds(CONNECT_SLEEP));
-    }
-
-    return event_handler->connectionRunning();
+    connection->getHandler().startBlockingLoop();
 }
 
 
 bool StorageRabbitMQ::updateChannel(ChannelPtr & channel)
 {
-    if (event_handler->connectionRunning())
+    if (connection->isConnected())
     {
-        channel = std::make_shared<AMQP::TcpChannel>(connection.get());
+        channel = connection->createChannel();
```

Comment:
Can connection be changed (closed for example) between calling `isConnected` and `createChannel`? BTW it would be better to create one additional method that checks if connection is running and return channel  under one mutex lock